### PR TITLE
SC2: Restrict allow Orphan to missions that already require that

### PR DESCRIPTION
**Pull request by Ziktofel**
## What is this fixing or adding?
Restricts setting allow orphan item flag only if a mission actually requiring that is rolled
Thus just having Liberation Day won't allow Medic upgrades as the player doesn't need that and Medics aren't even available in that mission

## How was this tested?
Unit Tests

## If this makes graphical changes, please attach screenshots.
No

### CODE REVIEW

File: worlds/sc2/__init__.py [hunk 1]
```diff
@@ -654,8 +654,7 @@ def flag_mission_based_item_excludes(world: SC2World, item_list: List[FilterItem
 def flag_allowed_orphan_items(world: SC2World, item_list: List[FilterItem]) -> None:
     """Adds the `Allowed_Orphan` flag to items that shouldn't be filtered with their parents, like combat shield"""
     missions = get_all_missions(world.custom_mission_order)
-    terran_nobuild_missions = any((MissionFlag.Terran|MissionFlag.NoBuild) in mission.flags and mission.campaign != SC2Campaign.NCO for mission in missions)
-    if terran_nobuild_missions:
+    if SC2Mission.PIERCING_OF_THE_SHROUD in missions:
```
Comment (MatthewMarinets) [inline worlds/sc2/__init__.py:657]: These orphan upgrades may also be used for the openings of The Dig, Engine of Destruction, and in Belly of the Beast. For BotB and The Dig, they are not required on advanced tactics; EoD and PtS can require them on advanced, though.

So this check should be
```python
have_terran_orphan_item_sections = (
    SC2Mission.PIERCING_OF_THE_SHROUD in missions
    or SC2Mission.ENGINE_OF_DESTRUCTION in missions
    or (world.options.required_tactics.value == RequiredTactics.option_standard
        and (
            SC2Mission.THE_DIG in missions
            or SC2Mission.BELLY_OF_THE_BEAST in missions
        )
    )
)
if have_terran_orphan_item_sections:
```
Edit: Thanks, Snarky, for spotting that I had the advanced logic flag on backwards. Fixed!

File: worlds/sc2/__init__.py [hunk 2]
```diff
@@ -665,7 +664,11 @@ def flag_allowed_orphan_items(world: SC2World, item_list: List[FilterItem]) -> N
     # These rules only trigger on Standard tactics
     if SC2Mission.BELLY_OF_THE_BEAST in missions and world.options.required_tactics == RequiredTactics.option_standard:
         for item in item_list:
-            if item.name in (item_names.FIREBAT_NANO_PROJECTORS, item_names.FIREBAT_NANO_PROJECTORS, item_names.FIREBAT_PROGRESSIVE_STIMPACK):
+            if item.name in (
+                    item_names.MARINE_COMBAT_SHIELD, item_names.MARINE_PROGRESSIVE_STIMPACK, item_names.MARINE_MAGRAIL_MUNITIONS,
+                    item_names.MEDIC_STABILIZER_MEDPACKS, item_names.MEDIC_NANO_PROJECTOR, item_names.MARINE_LASER_TARGETING_SYSTEM,
+                    item_names.FIREBAT_NANO_PROJECTORS, item_names.FIREBAT_NANO_PROJECTORS, item_names.FIREBAT_PROGRESSIVE_STIMPACK
```
Comment (MatthewMarinets) [inline worlds/sc2/__init__.py]: Medic nano projector is not in any of the logic rules and in fact isn't even a progression item.

The old purpose of this logic was to throw these items in the pool, even if they might be culled later for space, just to provide more variety for no-builds; marine/medic items are useful on every terran no-build (except medic ones on Liberation Day). Note they even affect Raynor's units on Breakout.

As such, I'm not sure the main thrust of this change is even necessary, as it only reduces the opportunity for variety on no-builds when these units are culled/excluded. It makes sense to add The Dig and Engine of Destruction to the list of missions to check; but otherwise I don't see the point of narrowing the use of this rule when it already had a decently tight fit on what players would likely want. If we really want to go through with making this filter a minimal fit to satisfy logic (which I'm not advocating for), then medic nano projectors should be taken off this list. Such a minimal fit probably isn't even necessary anymore because of the logic escape hatch on culling; so it would probably be more effective to just delete this whole function for an absolute minimal fit.

Edit: Actually, the old filter probably picks up With Friends Like These, where the items are not useful. That's a bug that should probably be fixed.

File: worlds/sc2/__init__.py (hunk 1 above)
Comment (Snarkie) [reply worlds/sc2/__init__.py:657]: From checking all the Terran missions with no-build portions, this should be correct. After the fix, these would cover all missions where the items can contribute to logic.

File: worlds/sc2/__init__.py (hunk 2 above)
Comment (Snarkie) [reply worlds/sc2/__init__.py]: I would go for the minimal fit option mentioned here. The situation is allowing items to be added to the pool when the parent unit for said item is NOT in the pool. So you only ever benefit from these items if you happen to get marines/medics as starting units, and as soon as you lose them, the items are dead weight. Allowing anything more than the minimum to fulfill logic seems like a waste to me. 

The variety argument is fair, but I would argue we can leave that to seeds that actually roll the parent item.

Either way, I agree that missions like WFLT where you do not get any marines/medics should not be considered.

**Comment by Ziktofel** [comment]
I'll add some units tests to cover missions like The Dig

File: worlds/sc2/__init__.py (hunk 2 above)
Comment (Ziktofel) [reply worlds/sc2/__init__.py]: Fixed the item list

File: worlds/sc2/__init__.py (hunk 1 above)
Comment (Ziktofel) [reply worlds/sc2/__init__.py:657]: Build missions needn't be on the list, see the newly added unit test. IIRC this is actually needed if you have only no-build missions of the given race, that have unit upgrade logic

File: worlds/sc2/__init__.py (hunk 2 above)
Comment (Ziktofel) [reply worlds/sc2/__init__.py]: And I'd leave fine-tuning for the next cycle. Now this is to just fix a clear bug

File: worlds/sc2/test/test_generation.py [hunk 3]
```diff
@@ -1226,3 +1227,36 @@ def test_exclude_overpowered_items_and_not_allow_unit_nerfs(self) -> None:
 
         # A unit nerf happens due to excluding OP items
         self.assertNotIn(item_names.MOTHERSHIP_INTEGRATED_POWER, starting_inventory)
+
+    def test_orphan_behavior_for_build_missions(self) -> None:
```
Comment (MatthewMarinets) [inline worlds/sc2/test/test_generation.py]: This test case passes without your changes. It also passes if you just entirely comment out the body of `flag_allowed_orphan_items()` so nothing is marked orphan. As such, I'm still unsure what the goal of this PR changelist is.

The `AllowedOrphan` flag was originally added for two reasons:
1. Allow logic to pass on no-builds with certain units culled/excluded. This is no longer necessary with the logic "escape hatch" mechanism on culling, which will refuse to exclude items necessary for a logic rule.
2. Promote variability in no-build missions so they're not the same every playthrough. Marines and medics show up on basically every terran no-build (except WFLT and medics aren't on Liberation Day). Player feedback explicitly listed firebat upgrades as good on Belly of the Beast.

I'm unsure why we're reducing the scope of the flag now when it hurts goal 2, but doesn't entirely clean up the code. If the goal is code / complexity cleanup, then nothing should be flagged as orphaned ~~(and I'd also like to see player feedback requesting this system be removed, as player feedback is what brought it into existence)~~. If the goal is to still promote variability, I'm unsure why the scope is being reduced.

Edit: I can't track down the original discussion around what exactly was marked as AllowOrphan, so it may not have come from player feedback. That means the only real reason I see to keep that flag around is for future work around option error handling as we've been discussing in the discord. For that use-case, we can just comment out the code or even remove the function entirely and document what items will get the flag in future under a github issue.

For the non-sc2 people looking for context on this future work, the problem being solved is that right now the sc2 generator is allowed to selectively ignore the user's exclusions if exclusion+random culling doesn't meet all logic rules for missions (which are also randomly selected). This error resolution strategy is what's making the allow orphan flag essentially a no-op. In future we want to support different error resolution strategies, such as excluding the offending locations/missions, or simply failing generation and printing a reasonable error message that allows the user to fix their yaml. Changing the strategy would mean the AllowOrphan flag would be necessary again.

Comment (Ziktofel) [reply worlds/sc2/test/test_generation.py]: Commenting that step out fails two following tests:
```
test_excluding_all_protoss_build_missions_excludes_protoss_units
test_excluding_all_terran_build_missions_excludes_all_terran_units
```
Both lead to generation failures.
Thus the behavior (actually flagging `AllowedOrphan`) is needed for nobuilds (but build missions are safe)

Comment (Ziktofel) [reply worlds/sc2/test/test_generation.py]: In a future dev cycle we can think about using `AllowedOrphan` more further but this is a bugfix as the problem was that having Liberation Day as the only Terran mission allowed you to have Medic upgrades while you never see any Medic in your game

Comment (MatthewMarinets) [reply worlds/sc2/test/test_generation.py]: Okay, looks like I was just getting tripped up by all the behaviour being covered by other test cases, while this test case covers something else. I'd argue that means this test is misnamed, as it has nothing to do with the orphan item flag -- maybe `test_terran_nobuild_sections_get_marine_medic_upgrades_with_units_excluded()`?

**Review by MatthewMarinets — APPROVED** [review]
It seems I was tripped up by this PR changing a little bit more than expected to fix the bug, but it should fix the bug and all the other changed behaviour is very edge-case anyways.

The added test case doesn't actually test the changed code, but instead displays that other code handles what Snarkie and I were suggesting for better coverage. I'd still prefer to see it renamed to something more descriptive, but it does its job.

As for changed behaviour not strictly necessary to fix the bug:
* When marine/medic are culled or excluded
* When Liberation Day, Ghost of a Chance, or Breakout generate in the mission order
* Old behaviour: some Marine or Medic upgrades could generate as part of the mission order
* New behaviour: Marine Medic upgrades are also culled

Liberation Day + Medic items was the bug being fixed here, the rest is a collateral change. The change seems reasonable enough; we can further change it if we get player feedback and it's edge-case enough that I don't expect anyone would be negatively impacted by it for many years.

File: worlds/sc2/test/test_generation.py (hunk 3 above)
Comment (Ziktofel) [reply worlds/sc2/test/test_generation.py]: Name updated
